### DISCUSSION TITLE
Operator overloading for the "with" operator

### DISCUSSION THREAD

**Discussion by khodaie** [dotnet/csharplang#6711] [Language Ideas]
The `with` expression is introduced in C# 9. As I know, it only works for the `record` and `struct` types. But I think by making it overloadable, it can be used in any classes:

````csharp
public class Person
{
    private int _internalField;

    public int Id { get; set;}
    public int Name { get; set; }
   
    public static Person operator with (Person person) => new Person { Id = person.Id, Name = person.Name, _internalField = person._internalField + 1 };
}

...

var person = new Person {...};
var newPerson = person with {Id = 2};
````

**Comment by pinkfloydx33** [comment]
FYI, all structs (records or otherwise) support `with` as of C#10. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQG4TIAEAZsXQLx0wCmA7nQGID2/OgG86AIVotmAXwDc1Wo0KtGzLlAQALEeIgAvFgGY6civmIBOABQAiBoQB0E5C2EAGaXCVP9r4tJsvex9kIMcJPQBKeXI6OLoKAGcENABjBD5BEQp4unxjWAznHQBzDgRZOkTyyukE2Pj8ukLdPVKaqo66ijqgA==), [Docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs#allow-with-expression-on-structs)

**Reply by khodaie** [reply]
> FYI, all structs (records or otherwise) support `with` as of C#10. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQG4TIAEAZsXQLx0wCmA7nQGID2/OgG86AIVotmAXwDc1Wo0KtGzLlAQALEeIgAvFgGY6civmIBOABQAiBoQB0E5C2EAGaXCVP9r4tJsvex9kIMcJPQBKeXI6OLoKAGcENABjBD5BEQp4unxjWAznHQBzDgRZOkTyyukE2Pj8ukLdPVKaqo66ijqgA==), [Docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs#allow-with-expression-on-structs)

Yes. Thanks. I edited the question.
